backend/compact-connect/lambdas/python/feature-flag/feature_flag_client.py

```diff
@@ -0,0 +1,629 @@
+        """
+        if not self._console_api_key:
+            raise FeatureFlagException(
+                'Console API key not configured. Required for management operations (create, update, delete).'
+            )
+
+        url = f'{STATSIG_API_BASE_URL}{endpoint}'
+        headers = {
+            'STATSIG-API-KEY': self._console_api_key,
+            'STATSIG-API-VERSION': STATSIG_API_VERSION,
+            'Content-Type': 'application/json',
+        }
+
+        try:
+            if method == 'GET':
+                response = requests.get(url, headers=headers, timeout=30)
+            elif method == 'POST':
+                response = requests.post(url, headers=headers, data=json.dumps(data), timeout=30)
+            elif method == 'PATCH':
+                response = requests.patch(url, headers=headers, data=json.dumps(data), timeout=30)
+            elif method == 'DELETE':
+                response = requests.delete(url, headers=headers, timeout=30)
+            else:
+                raise ValueError(f'Unsupported HTTP method: {method}')
+
+            return response
+
+        except requests.exceptions.RequestException as e:
+            raise FeatureFlagException(f'StatSig Console API request failed: {e}') from e
+
+    def upsert_flag(
+        self, flag_name: str, auto_enable: bool = False, custom_attributes: dict[str, Any] | None = None
+    ) -> dict[str, Any]:
+        """
+        Create or update a feature gate in StatSig.
+
+        Each environment has its own rule (e.g., 'test-rule', 'beta-rule', 'prod-rule').
+        - If auto_enable is False: passPercentage is set to 0 (disabled)
+        - If auto_enable is True: passPercentage is set to 100 (enabled) and custom attributes are applied
+
+        :param flag_name: Name of the feature gate
+        :param auto_enable: If True, enable the flag (passPercentage=100); if False, disable it (passPercentage=0)
+        :param custom_attributes: Optional custom attributes for targeting (only applied if auto_enable=True)
+        :return: Flag data (with 'id' field)
+        :raises FeatureFlagException: If operation fails
+        """
+        # Check if gate already exists
+        existing_gate = self.get_flag(flag_name)
+
+        if not existing_gate:
+            # Create new gate with environment-specific rule
+            return self._create_new_gate(flag_name, auto_enable, custom_attributes)
+
+        # Gate exists - check if environment rule exists
+        gate_id = existing_gate.get('id')
+        rule_name = f'{self.environment.lower()}-rule'
+        environment_rule = self._find_environment_rule(existing_gate, rule_name)
+
+        # we only set the environment rule if it doesn't already exist
+        # else we leave it alone to avoid overwriting manual changes
+        if not environment_rule:
+            updated_gate = self._add_environment_rule(existing_gate, auto_enable, custom_attributes)
```

Comment:
It took me a minute to understand what this was doing here, adding an environment rule, then updating the gate. I assumed, based on the `_add_environment_rule` name, that the gate would have been updated before the `_update_gate` call. Maybe call it `_add_environment_rule_to_gate_data` or something, for clarity? Or maybe actually call `self._update_gate()` from `_add_environment_rule()`?